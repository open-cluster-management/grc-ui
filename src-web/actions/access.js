/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

import {
  USER_ACCESS_SUCCESS, USER_ACCESS_FAILURE,
} from './index'
import GrcApolloClient from '../utils/client/apollo-client'

export const userAccessSuccess = access => ({
  type: USER_ACCESS_SUCCESS,
  access
})

export const userAccessFailure = err => ({ type: USER_ACCESS_FAILURE, err })

export const getUserAccessData = () => {
  return (dispatch) => {
    return GrcApolloClient.getUserAccess()
      .then(response => {
        if (response.errors) {
          return dispatch(userAccessFailure(response.errors[0]))
        }
        return dispatch(userAccessSuccess(response.data.items))
      })
      .catch(err => dispatch(userAccessFailure(err)))
  }
}
