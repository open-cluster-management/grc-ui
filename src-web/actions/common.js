/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
import _ from 'lodash'

import GrcApolloClient from '../../lib/client/apollo-client'
import { formatExpandablePolicies } from '../components/common/FormatTableData'
import {
  TABLE_PAGE_CHANGE, TABLE_SEARCH, TABLE_SORT, RESOURCE_RECEIVE_SUCCESS,
  REQUEST_STATUS, RESOURCE_RECEIVE_FAILURE, RESOURCE_REQUEST, RESOURCE_ADD,
  RESOURCE_MODIFY, RESOURCE_DELETE, RESOURCE_MUTATE, RESOURCE_MUTATE_SUCCESS,
  RESOURCE_MUTATE_FAILURE, AVAILABLE_FILTER_UPDATE, ACTIVE_FILTER_UPDATE,
  SECONDARY_HEADER_UPDATE, MODAL_UPDATE, POST_REQUEST, POST_RECEIVE_SUCCESS,
  POST_RECEIVE_FAILURE, PUT_REQUEST, PUT_RECEIVE_SUCCESS, PUT_RECEIVE_FAILURE,
  PATCH_REQUEST, PATCH_RECEIVE_SUCCESS, PATCH_RECEIVE_FAILURE, DEL_REQUEST,
  DEL_RECEIVE_SUCCESS, DEL_RECEIVE_FAILURE, CLEAR_REQUEST_STATUS, RESOURCE_RESET,
} from './index'

export const changeTablePage = ({page, pageSize}, resourceType) => ({
  type: TABLE_PAGE_CHANGE,
  page,
  pageSize,
  resourceType
})

export const searchTable = (search, resourceType) => ({
  type: TABLE_SEARCH,
  search,
  resourceType
})

export const sortTable = (sortDirection, sortColumn, resourceType) => ({
  type: TABLE_SORT,
  sortDirection,
  sortColumn,
  resourceType
})

export const receiveResourceSuccess = (response, resourceType) => ({
  type: RESOURCE_RECEIVE_SUCCESS,
  status: REQUEST_STATUS.DONE,
  items: response.items,
  resourceVersion: _.get(response, 'metadata.resourceVersion'), //only supported on k8s resoruces
  resourceType
})

export const receiveResourceError = (err, resourceType) => ({
  type: RESOURCE_RECEIVE_FAILURE,
  status: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const requestResource = (resourceType) => ({
  type: RESOURCE_REQUEST,
  status: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const addResource = (item, resourceType) => ({
  type: RESOURCE_ADD,
  resourceType: item.kind || resourceType,
  item
})

export const modifyResource = (item, resourceType) => ({
  type: RESOURCE_MODIFY,
  resourceType: item.kind || resourceType,
  item
})

export const deleteResource = (item, resourceType) => ({
  type: RESOURCE_DELETE,
  resourceType: item.kind || resourceType,
  item
})

export const mutateResource = (resourceType, resourceName) => ({
  type: RESOURCE_MUTATE,
  resourceName,
  resourceType,
})

export const mutateResourceSuccess = (resourceType, resourceName) => ({
  type: RESOURCE_MUTATE_SUCCESS,
  resourceName,
  resourceType,
})

export const mutateResourceFailure = (resourceType, error) => ({
  type: RESOURCE_MUTATE_FAILURE,
  postStatus: REQUEST_STATUS.ERROR,
  err: { error },
  resourceType,
})

export const fetchResources = (resourceType, vars) => {
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.get(resourceType, vars)
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(formatExpandablePolicies(response.data.items))}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const fetchSingleResource = (resourceType, args) => {
  console.log(resourceType)
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.getResource(resourceType, args)
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(response.data)}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const fetchResource = (resourceType, namespace, name) => {
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.getResource(resourceType, {namespace, name})
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(response.data.items)}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const editResource = (resourceType, namespace, name, body, selfLink, resourcePath) => (dispatch => {
  dispatch(putResource(resourceType))
  return GrcApolloClient.updateResource(resourceType.name, namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePutError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-edit'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePutResource(response, resourceType))
    })
})

export const disableResource = (resourceType, namespace, name, body, selfLink, resourcePath) => (dispatch => {
  dispatch(patchResource(resourceType))
  return GrcApolloClient.updateResource(resourceType.name, namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePatchError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-disable'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePatchResource(response, resourceType))
    })
})

export const enforcResource = (resourceType, namespace, name, body, selfLink, resourcePath) => (dispatch => {
  dispatch(patchResource(resourceType))
  return GrcApolloClient.updateResource(resourceType.name, namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePatchError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-enforce'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePatchResource(response, resourceType))
    })
})

export const removeResource = (resourceType, vars) => async dispatch => {
  dispatch(delResource(resourceType))
  try {
    const response = await GrcApolloClient.remove(vars)
    if (response.errors) {
      return dispatch(receiveDelError(response.errors, resourceType))
    } else {
      dispatch(receiveDelResource(response, resourceType, vars))
    }
    dispatch(fetchResources(resourceType))
  } catch (err) {
    return dispatch(receiveDelError(err, resourceType))
  }
}

export const updateActiveFilters = (activeFilters) => ({
  type: ACTIVE_FILTER_UPDATE,
  activeFilters,
})

export const updateAvailableFilters = (availableFilters) => ({
  type: AVAILABLE_FILTER_UPDATE,
  availableFilters
})

export const updateSecondaryHeader = (title, tabs, breadcrumbItems, links, description, information) => ({
  type: SECONDARY_HEADER_UPDATE,
  title,
  tabs,
  breadcrumbItems,
  links,
  description,
  information
})

export const updateModal = (data) => ({
  type: MODAL_UPDATE,
  data
})

export const postResource = (resourceType) => ({ // TODO: Consider renaming
  type: POST_REQUEST,
  postStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePostResource = (item, resourceType) => ({
  type: POST_RECEIVE_SUCCESS,
  postStatus: REQUEST_STATUS.DONE,
  resourceType: item.kind || resourceType,
  item
})

export const receivePostError = (err, resourceType) => ({
  type: POST_RECEIVE_FAILURE,
  postStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const putResource = (resourceType) => ({ // TODO: Consider renaming
  type: PUT_REQUEST,
  putStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePutResource = (item, resourceType) => {
  return ({
    type: PUT_RECEIVE_SUCCESS,
    putStatus: REQUEST_STATUS.DONE,
    resourceType: item.kind || resourceType,
    item
  })
}

export const receivePutError = (err, resourceType) => ({
  type: PUT_RECEIVE_FAILURE,
  putStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const patchResource = (resourceType) => ({ // TODO: Consider renaming
  type: PATCH_REQUEST,
  patchStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePatchResource = (item, resourceType) => {
  return ({
    type: PATCH_RECEIVE_SUCCESS,
    patchStatus: REQUEST_STATUS.DONE,
    resourceType: item.kind || resourceType,
    item
  })
}

export const receivePatchError = (err, resourceType) => ({
  type: PATCH_RECEIVE_FAILURE,
  patchStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const delResource = (resourceType) => ({ // TODO: Consider renaming
  type: DEL_REQUEST,
  delStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receiveDelResource = (item, resourceType, resource) => ({
  type: DEL_RECEIVE_SUCCESS,
  delStatus: REQUEST_STATUS.DONE,
  resourceType: item.kind || resourceType,
  item,
  resource
})

export const receiveDelError = (err, resourceType) => ({
  type: DEL_RECEIVE_FAILURE,
  delStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const clearRequestStatus = (resourceType) => ({
  type: CLEAR_REQUEST_STATUS,
  resourceType: resourceType
})

export const resetResource = (resourceType) => ({
  type: RESOURCE_RESET,
  resourceType: resourceType
})

export const createResources = (resourceType, resourceJson) => {
  return (dispatch) => {
    dispatch(mutateResource(resourceType))
    return GrcApolloClient.createResources(resourceJson)
      .then(result => {
        if (result.data.createResources.errors && result.data.createResources.errors.length > 0){
          dispatch(mutateResourceFailure(resourceType, result.data.createResources.errors[0]))
        } else {
          dispatch(mutateResourceSuccess(resourceType))
        }
        return result
      })
  }
}

export const createAndUpdateResources = (resourceTypes, createList, updateList) => {
  return (dispatch) => {
    resourceTypes.forEach((resourceType) => {
      dispatch(mutateResource(resourceType))
    })
    return GrcApolloClient.createAndUpdateResources(createList, updateList)
      .then(result => {
        const errors = {
          Policy: {
            resourceType: 'Policy',
            error: '',
          },
          PlacementRule: {
            resourceType: 'PlacementRule',
            error: '',
          },
          PlacementBinding: {
            resourceType: 'PlacementBinding',
            error: '',
          }
        }
        if (_.get(result, 'data.createAndUpdateResources.create.errors') &&
          _.get(result, 'data.createAndUpdateResources.create.errors.length') > 0){
          _.get(result, 'data.createAndUpdateResources.create.errors', []).forEach((error) => {
            errors[error.kind].error = error.message
          })
        }
        if (_.get(result, 'data.createAndUpdateResources.update.errors') &&
          _.get(result, 'data.createAndUpdateResources.update.errors.length') > 0){
          _.get(result, 'data.createAndUpdateResources.update.errors', []).forEach((error) => {
            errors[error.kind].error = error.message
          })
        }
        let errored = false
        Object.keys(errors).forEach((key) => {
          const resp = errors[key]
          if (resp.error !== '') {
            errored = true
            dispatch(mutateResourceFailure(resp.resourceType, { message: resp.error }))
          }
        })
        if (!errored) {
          resourceTypes.forEach((resourceType) => {
            dispatch(mutateResourceSuccess(resourceType))
          })
        }
        return result
      })
  }
}

export const createResource = (resourceType, variables) => {
  return (dispatch) => {
    dispatch(postResource(resourceType))
    return GrcApolloClient.createResource(resourceType, variables)
      .then(response => {
        if (response.errors) {
          return dispatch(receivePostError(response.errors[0], resourceType))
        }

        return dispatch(receivePostResource(_.cloneDeep(response.data.setHelmRepo), resourceType))
      })
      .catch(err => dispatch(receivePostError(err, resourceType)))
  }
}
