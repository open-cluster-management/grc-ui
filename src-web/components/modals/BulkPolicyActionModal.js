/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import {
	AcmAlert,
	AcmButton,
	AcmModal,
	AcmTable,
} from '@open-cluster-management/ui-components'
import { Spinner, ButtonVariant } from '@patternfly/react-core'
import msgs from '../../nls/platform.properties'
import { withRouter } from 'react-router-dom'
import { REQUEST_STATUS } from '../../actions/index'
import {
	bulkPolicyActions,
	clearRequestStatus,
	receivePatchError,
	updateModal,
} from '../../actions/common'

const policyTableColumns = (modalType) => {
	switch (modalType) {
		case 'bulk-policy-action-inform':
		case 'bulk-policy-action-enforce':
			return [
				{
					header: 'Policy name',
					sort: 'name.rawData',
					cell: 'name.rawData',
					search: 'name.rawData',
				},
				{
					header: 'Source',
					sort: 'source.text',
					cell: 'source.text',
				},
				{
					header: 'Remediation',
					sort: 'remediation.title',
					cell: 'remediation.title',
				},
			]
		case 'bulk-policy-action-enable':
		case 'bulk-policy-action-disable':
			return [
				{
					header: 'Policy name',
					sort: 'name.rawData',
					cell: 'name.rawData',
					search: 'name.rawData',
				},
				{
					header: 'Source',
					sort: 'source.text',
					cell: 'source.text',
				},
				{
					header: 'Status',
					sort: 'status.text',
					cell: 'status.text',
				},
			]
		/* istanbul ignore next */
		default:
			break
	}
}

export class BulkPolicyActionModal extends React.Component {
	constructor(props) {
		super(props)
		this.handleSubmitClick = this.handleSubmitClick.bind(this)
		this.handleCloseClick = this.handleCloseClick.bind(this)
	}

	handleSubmitClick() {
		const {
			type: modalType,
			handleSubmit,
			data = [],
		} = this.props
		switch (modalType) {
			case 'bulk-policy-action-disable':
				handleSubmit(
					data,
					true,
					'/spec/disabled',
					modalType,
					bulkPolicyActions
				)
				break
			case 'bulk-policy-action-enable':
				handleSubmit(
					data,
					false,
					'/spec/disabled',
					modalType,
					bulkPolicyActions
				)
				break
			case 'bulk-policy-action-enforce':
				handleSubmit(
					data,
					'enforce',
					'/spec/remediationAction',
					modalType,
					bulkPolicyActions
				)
				break
			case 'bulk-policy-action-inform':
				handleSubmit(
					data,
					'inform',
					'/spec/remediationAction',
					modalType,
					bulkPolicyActions
				)
				break
			default:
				break
		}
	}

	handleCloseClick() {
		const { type: modalType, handleClose } = this.props
		handleClose(modalType)
	}

	render() {
		const {
			actionType,
			type: modalType,
			data,
			label,
			locale,
			open,
			reqErrorMsg,
			reqStatus,
		} = this.props
		return (
			<div>
				{reqStatus === REQUEST_STATUS.IN_PROGRESS && (
					<Spinner className="patternfly-spinner" />
				)}
				<AcmModal
					variant="medium"
					id={`bulk-${actionType}-policy-modal'`}
					isOpen={open}
					showClose={true}
					onClose={this.handleCloseClick}
					title={
						data.length > 1
							? msgs.get(label.heading, [data.length], locale)
							: msgs.get(
									`modal.${actionType}-hcmcompliance.heading`,
									locale
							)
					}
					actions={[
						<AcmButton
							key="confirm"
							variant={ButtonVariant.primary}
							onClick={this.handleSubmitClick}
						>
							{msgs.get(label.primaryBtn, locale)}
						</AcmButton>,
						<AcmButton
							key="cancel"
							variant={ButtonVariant.link}
							onClick={this.handleCloseClick}
						>
							{msgs.get('modal.button.cancel', locale)}
						</AcmButton>,
					]}
				>
					{reqStatus === REQUEST_STATUS.ERROR && (
						<AcmAlert
							isInline={true}
							noClose={true}
							variant="warning"
							title={
								reqErrorMsg ||
								msgs.get('error.default.description', locale)
							}
						/>
					)}
					<p style={{ marginBottom: '.5rem' }}>
						{msgs.get(
							`modal.actions.bulk.${actionType}.description`,
							locale
						)}
					</p>
					{data.some(
						(policy) => policy.source.text.toLowerCase() !== 'local'
					) && (
						<AcmAlert
							isInline={true}
							noClose={true}
							variant="default"
							title={msgs.get(
								'modal.actions.bulk.external.alert.title',
								[
									data.filter(
										(policy) =>
											policy.source.text.toLowerCase() !==
											'local'
									).length,
								],
								locale
							)}
							message={msgs.get(
								'modal.actions.bulk.external.alert.message',
								locale
							)}
						/>
					)}
					<AcmTable
						plural="policy actions"
						items={data}
						searchPlaceholder={msgs.get(
							'modal.actions.bulk.table.search.placeholder',
							locale
						)}
						columns={policyTableColumns(modalType)}
						keyFn={(item) => item.uid.toString()}
						gridBreakPoint=""
					/>
				</AcmModal>
			</div>
		)
	}
}

BulkPolicyActionModal.propTypes = {
	actionType: PropTypes.string,
	data: PropTypes.array,
	handleClose: PropTypes.func,
	handleSubmit: PropTypes.func,
	label: PropTypes.shape({
		heading: PropTypes.string,
		label: PropTypes.string,
		primaryBtn: PropTypes.string,
	}),
	locale: PropTypes.string,
	open: PropTypes.bool,
	reqErrorMsg: PropTypes.string,
	reqStatus: PropTypes.string,
	type: PropTypes.string,
}

/* istanbul ignore next */
const mapStateToProps = (state) => {
	return state.modal
}

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => {
	return {
		handleSubmit: (
			policies,
			newData,
			resourcePath,
			modalType,
			dispatchFn
		) => {
			dispatch(dispatchFn(policies, newData, resourcePath, modalType))
		},
		handleClose: (modalType) => {
			dispatch(clearRequestStatus())
			dispatch(updateModal({ open: false, type: modalType }))
		},
		receivePatchError: (resourceType, err) => {
			dispatch(receivePatchError(err, resourceType))
		},
	}
}

export default withRouter(
	connect(mapStateToProps, mapDispatchToProps)(BulkPolicyActionModal)
)
