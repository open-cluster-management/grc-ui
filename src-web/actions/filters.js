/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

import _ from 'lodash'

import GrcApolloClient from '../utils/client/apollo-client'
import { receiveResourceError, requestResource}  from './common'
import { RESOURCE_TYPES } from '../utils/constants'
import {
  REQUEST_STATUS, RESOURCE_FILTERS_RECEIVE_SUCCESS, RESOURCE_FILTERS_UPDATE,
} from './index'

export const STRING_SPLITTER = '='

export  const combineFilters = (selectedFilters) => {
  const tempObject = []
  if (selectedFilters && selectedFilters.length > 0) {
    selectedFilters.forEach(item => {
      if (item && item.type) {
        const { key, value, type } = item
        tempObject.push({key, value, type})
      } else {
        //TODO support customize tags
      }
    })
  }
  return {filter: {resourceFilter: tempObject}}
}

export const fetchFilters = (inputType) => {
  const resourceType = RESOURCE_TYPES.HCM_FILTER_LIST
  return (dispatch) => {
    if (inputType && inputType.filter) {
      dispatch(requestResource(resourceType))
      return GrcApolloClient.get(resourceType)
        .then(response => {
          if (response.errors) {
            return dispatch(receiveResourceError(response.errors[0], resourceType))
          }
          return dispatch(receiveFiltersSuccess({
            clusterLabels: _.get(response, 'data.filters.clusterLabels'),
            clusterNames: _.get(response, 'data.filters.clusterNames'),
          }, resourceType))
        })
        .catch(err => dispatch(receiveResourceError(err, resourceType)))
    }
    return undefined
  }
}

export const receiveFiltersSuccess = (response, resourceType) => ({
  type: RESOURCE_FILTERS_RECEIVE_SUCCESS,
  status: REQUEST_STATUS.DONE,
  filters: {
    clusterLabels: response.clusterLabels || [],
    clusterNames: response.clusterNames || [],
  },
  resourceType
})

export const updateResourceFilters = (resourceType, selectedFilters) => ({
  type: RESOURCE_FILTERS_UPDATE,
  resourceName: resourceType.name,
  selectedFilters,
})
