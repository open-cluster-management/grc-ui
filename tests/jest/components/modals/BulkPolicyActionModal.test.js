/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import { BulkPolicyActionModal } from '../../../../src-web/components/modals/BulkPolicyActionModal'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { AcmAlert } from '@stolostron/ui-components'
import { REQUEST_STATUS } from '../../../../src-web/actions/index'

describe('PolicyActionModal component', () => {
    it('Modal successfully renders and performs inform bulk action', () => {
        const onHandleClose = jest.fn()
        const onHandleSubmit = jest.fn()
		const wrapper = mount(
			<BulkPolicyActionModal
                actionType={'inform'}
                data={[
                    {
                        name: {
                            rawData: 'test-external-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            rawData: {
                                gitPath: 'stable',
                                gitBranch: 'main',
                                type: 'GitHub',
                                pathname: 'https://github.com/stolostron/policy-collection.git'
                            },
                            text: 'Git',
                        },
                        status: {
                            rawData: false,
                            text: 'Enabled',
                            title: 'Enabled',
                        },
                        uid: '1',
                    },
                    {
                        name: {
                            rawData: 'test-local-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            text: 'local',
                        },
                        status: {
                            rawData: false,
                            text: 'Disabled',
                            title: 'Disabled',
                        },
                        uid: '2',
                    },
                ]}
                handleClose={onHandleClose}
                handleSubmit={onHandleSubmit}
                label={{
                    primaryBtn: 'modal.actions.bulk.inform.primaryBtn',
                    label: 'label',
                    heading: 'modal.actions.bulk.inform.heading'
                }}
                open={true}
                type={'bulk-policy-action-inform'}
			/>
		)
        wrapper.find('.pf-m-primary').simulate('click')
        expect(onHandleSubmit).toHaveBeenCalled()
    })
    it('Modal successfully renders and performs enforce bulk action', () => {
        const onHandleClose = jest.fn()
        const onHandleSubmit = jest.fn()
		const wrapper = mount(
			<BulkPolicyActionModal
                actionType={'enforce'}
                data={[
                    {
                        name: {
                            rawData: 'test-external-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            rawData: {
                                gitPath: 'stable',
                                gitBranch: 'main',
                                type: 'GitHub',
                                pathname: 'https://github.com/stolostron/policy-collection.git'
                            },
                            text: 'Git',
                        },
                        status: {
                            rawData: false,
                            text: 'Enabled',
                            title: 'Enabled',
                        },
                        uid: '1',
                    },
                    {
                        name: {
                            rawData: 'test-local-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            text: 'local',
                        },
                        status: {
                            rawData: false,
                            text: 'Disabled',
                            title: 'Disabled',
                        },
                        uid: '2',
                    },
                ]}
                handleClose={onHandleClose}
                handleSubmit={onHandleSubmit}
                label={{
                    primaryBtn: 'modal.actions.bulk.enforce.primaryBtn',
                    label: 'label',
                    heading: 'modal.actions.bulk.enforce.heading'
                }}
                open={true}
                type={'bulk-policy-action-enforce'}
			/>
		)
        wrapper.find('.pf-m-primary').simulate('click')
        expect(onHandleSubmit).toHaveBeenCalled()
    })
    it('Modal successfully renders and performs enable bulk action', () => {
        const onHandleClose = jest.fn()
        const onHandleSubmit = jest.fn()
		const wrapper = mount(
			<BulkPolicyActionModal
                actionType={'enable'}
                data={[
                    {
                        name: {
                            rawData: 'test-external-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            rawData: {
                                gitPath: 'stable',
                                gitBranch: 'main',
                                type: 'GitHub',
                                pathname: 'https://github.com/stolostron/policy-collection.git'
                            },
                            text: 'Git',
                        },
                        status: {
                            rawData: false,
                            text: 'Enabled',
                            title: 'Enabled',
                        },
                        uid: '1',
                    },
                    {
                        name: {
                            rawData: 'test-local-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            text: 'local',
                        },
                        status: {
                            rawData: false,
                            text: 'Disabled',
                            title: 'Disabled',
                        },
                        uid: '2',
                    },
                ]}
                handleClose={onHandleClose}
                handleSubmit={onHandleSubmit}
                label={{
                    primaryBtn: 'modal.actions.bulk.enable.primaryBtn',
                    label: 'label',
                    heading: 'modal.actions.bulk.enable.heading'
                }}
                open={true}
                type={'bulk-policy-action-enable'}
			/>
		)
        wrapper.find('.pf-m-primary').simulate('click')
        expect(onHandleSubmit).toHaveBeenCalled()
    })
    it('Modal successfully renders and performs disable bulk action', () => {
        const onHandleClose = jest.fn()
        const onHandleSubmit = jest.fn()
		const wrapper = mount(
			<BulkPolicyActionModal
                actionType={'disable'}
                data={[
                    {
                        name: {
                            rawData: 'test-external-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            rawData: {
                                gitPath: 'stable',
                                gitBranch: 'main',
                                type: 'GitHub',
                                pathname: 'https://github.com/stolostron/policy-collection.git'
                            },
                            text: 'Git',
                        },
                        status: {
                            rawData: false,
                            text: 'Enabled',
                            title: 'Enabled',
                        },
                        uid: '1',
                    },
                    {
                        name: {
                            rawData: 'test-local-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            text: 'local',
                        },
                        status: {
                            rawData: false,
                            text: 'Disabled',
                            title: 'Disabled',
                        },
                        uid: '2',
                    },
                ]}
                handleClose={onHandleClose}
                handleSubmit={onHandleSubmit}
                label={{
                    primaryBtn: 'modal.actions.bulk.disable.primaryBtn',
                    label: 'label',
                    heading: 'modal.actions.bulk.disable.heading'
                }}
                open={true}
                type={'bulk-policy-action-disable'}
			/>
		)
        wrapper.find('.pf-m-primary').simulate('click')
        expect(onHandleSubmit).toHaveBeenCalled()
    })
	it('Modal returns notification on error', () => {
        const component = shallow(
            <BulkPolicyActionModal
                actionType={'inform'}
                data={[
                    {
                        name: {
                            rawData: 'test-external-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'inform',
                            title: 'Inform',
                        },
                        source: {
                            rawData: {
                                gitPath: 'stable',
                                gitBranch: 'main',
                                type: 'GitHub',
                                pathname: 'https://github.com/stolostron/policy-collection.git'
                            },
                            text: 'Git',
                        },
                        status: {
                            rawData: false,
                            text: 'Enabled',
                            title: 'Enabled',
                        },
                    },
                    {
                        name: {
                            rawData: 'test-local-policy',
                        },
                        namespace: 'policies',
                        remediation: {
                            rawData: 'enforce',
                            title: 'Enforce',
                        },
                        source: {
                            text: 'local',
                        },
                        status: {
                            rawData: false,
                            text: 'Disabled',
                            title: 'Disabled',
                        },
                    },
                ]}
                handleClose={jest.fn()}
                handleSubmit={jest.fn()}
                label={{
                    primaryBtn: 'modal.actions.bulk.inform.primaryBtn',
                    label: 'label',
                    heading: 'modal.actions.bulk.inform.heading'
                }}
                open={true}
                type={'bulk-policy-action-inform'}

                locale={'en'}
                reqStatus={REQUEST_STATUS.ERROR}
                reqErrorMsg='There was an error.'
            />
        )
        expect(component.find(AcmAlert)).toHaveLength(2)
        expect(toJson(component)).toMatchSnapshot()
    })
})
