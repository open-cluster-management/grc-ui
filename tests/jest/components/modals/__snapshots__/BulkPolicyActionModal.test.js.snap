// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyActionModal component Modal returns notification on error 1`] = `
<div>
  <AcmModal
    actions={
      Array [
        <AcmButton
          onClick={[Function]}
          variant="primary"
        >
          Inform all
        </AcmButton>,
        <AcmButton
          onClick={[Function]}
          variant="link"
        >
          Cancel
        </AcmButton>,
      ]
    }
    id="bulk-inform-policy-modal'"
    isOpen={true}
    onClose={[Function]}
    showClose={true}
    title="Inform {0} policies?"
    variant="medium"
  >
    <AcmAlert
      isInline={true}
      noClose={true}
      title="There was an error."
      variant="warning"
    />
    <p
      style={
        Object {
          "marginBottom": ".5rem",
        }
      }
    >
      After you set the selected policies to inform, violations are reported and manual remediation is required.
    </p>
    <AcmAlert
      isInline={true}
      message="Any changes made here may be overridden by the content of an upstream repository."
      noClose={true}
      title="{0} policies are managed externally"
      variant="default"
    />
    <AcmTable
      columns={
        Array [
          Object {
            "cell": "name.rawData",
            "header": "Policy name",
            "search": "name.rawData",
            "sort": "name.rawData",
          },
          Object {
            "cell": "source.text",
            "header": "Source",
            "sort": "source.text",
          },
          Object {
            "cell": "remediation.title",
            "header": "Remediation",
            "sort": "remediation.title",
          },
        ]
      }
      gridBreakPoint=""
      items={
        Array [
          Object {
            "name": Object {
              "rawData": "test-external-policy",
            },
            "namespace": "policies",
            "remediation": Object {
              "rawData": "inform",
              "title": "Inform",
            },
            "source": Object {
              "rawData": Object {
                "gitBranch": "main",
                "gitPath": "stable",
                "pathname": "https://github.com/open-cluster-management/policy-collection.git",
                "type": "GitHub",
              },
              "text": "Git",
            },
            "status": Object {
              "rawData": false,
              "text": "Enabled",
              "title": "Enabled",
            },
          },
          Object {
            "name": Object {
              "rawData": "test-local-policy",
            },
            "namespace": "policies",
            "remediation": Object {
              "rawData": "enforce",
              "title": "Enforce",
            },
            "source": Object {
              "text": "local",
            },
            "status": Object {
              "rawData": false,
              "text": "Disabled",
              "title": "Disabled",
            },
          },
        ]
      }
      keyFn={[Function]}
      plural="policy actions"
      searchPlaceholder="Find by name"
    />
  </AcmModal>
</div>
`;
